File: solutions/Python/insert_interval_greedy.py
from typing import List

def insert(intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
    """Greedy approach: O(n) time, O(n) space"""
    result, i, new_start, new_end = [], 0, newInterval[0], newInterval[1]
    while i < len(intervals) and intervals[i][1] < new_start:
        result.append(intervals[i])
        i += 1
    while i < len(intervals) and intervals[i][0] <= new_end:
        new_start = min(new_start, intervals[i][0])
        new_end = max(new_end, intervals[i][1])
        i += 1
    result.append([new_start, new_end])
    while i < len(intervals):
        result.append(intervals[i])
        i += 1
    return result

File: solutions/Golang/insert_interval_greedy.go
package main

func insert(intervals [][]int, newInterval []int) [][]int {
    result := [][]int{}
    i, new_start, new_end := 0, newInterval[0], newInterval[1]
    for i < len(intervals) && intervals[i][1] < new_start {
        result = append(result, intervals[i])
        i++
    }
    for i < len(intervals) && intervals[i][0] <= new_end {
        if intervals[i][0] < new_start { new_start = intervals[i][0] }
        if intervals[i][1] > new_end { new_end = intervals[i][1] }
        i++
    }
    result = append(result, []int{new_start, new_end})
    for i < len(intervals) {
        result = append(result, intervals[i])
        i++
    }
    return result
}
